### PR TITLE
[FLINK-12413] [runtime] Implement ExecutionFailureHandler

### PR DESCRIPTION
**Pull request by zhuzhurk**
## What is the purpose of the change

*ExecutionFailureHandler is a component of new generation scheduling. See [FLINK-10429](https://issues.apache.org/jira/browse/FLINK-10429). It serves task failover handling by returning tasks to restart upon task failures. This PR is to implement ExecutionFailureHandler and related classes.*

## Brief change log

Detailed design doc can be found at [design](https://docs.google.com/document/d/1D4EP9i5rx9CTQ4-iRu6DVqVEG8KzIMu07Lk4zjxLy08/edit#)
  - *Implement ExecutionFailureHandler*
  - *Implement FailureHandlingResult*
  - *Add interface RestartBackoffTimeStrategy and RestartBackoffTimeStrategy.Factory*
  - *Add interface FailoverStrategy.Factory*

## Verifying this change

This change added tests and can be verified as follows:

*(example:)*
  - *UT for ExecutionFailureHandler*
  - *UT for FailureHandlingResult*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+	// ------------------------------------------------------------------------
+	//  utilities
+	// ------------------------------------------------------------------------
+
+	private ExecutionFailureHandler createExecutionFailureHandler() {
+		FailoverTopology topology = new TestFailoverTopology.Builder().build();
+
+		FailoverStrategy.Factory failoverStrategyFactory = mock(FailoverStrategy.Factory.class);
+		when(failoverStrategyFactory.create(any(FailoverTopology.class))).thenReturn(mock(FailoverStrategy.class));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java]: Please create a proper implementation of the FailoverStrategy. The factory can then be a simple lambda `topology -> new MyFailoverStrategy`.

Mocking things usually just leads to problems down the line.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+				failoverStrategy.getTasksNeedingRestart(failedTask, cause),
+				restartBackoffTimeStrategy.getBackoffTime());
+		} else {
+			return new FailureHandlingResult(
+				new JobException("Failed task restarting is suppressed by " + restartBackoffTimeStrategy, cause));
+		}
+	}
+
+	@VisibleForTesting
+	FailoverStrategy getFailoverStrategy() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: this method (and `getRestartBackoffTimeStrategy`) seem unnecessary; any test using this class must pass factories for the strategies, and thus can already access the actual strategy instance.
This could be as simple as:
```
FailoverStrategy failoverStrategy = new MyFailoverStrategy(failoverTopology);
RestartBackoffTimeStrategy restartStrategy = new MyRestartBackoffTimeStrategy();

ExecutionFailureHandler handler = new ExecutionFailureHandler(
	failoverTopology,
	topology -> failoverStrategy,
	() -> restartStrategy);
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ */
+public class FailureHandlingResultTest extends TestLogger {
+
+	/**
+	 * Tests normal FailureHandlingResult.
+	 */
+	@Test
+	public void testNormalFailureHandlingResult() throws Exception {
+		// create a normal FailureHandlingResult
+		Set<ExecutionVertexID> tasks = new HashSet<>();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java]: let's add at least 1 value to this set, this test would currently pass even if the result would always return an empty set of vertices.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	/**
+	 * Tests the case that task restarting is accepted.
+	 */
+	@Test
+	public void testNormalFailureHandling() throws Exception {
+		ExecutionFailureHandler executionFailureHandler = createExecutionFailureHandler();
+		FailoverStrategy failoverStrategy = executionFailureHandler.getFailoverStrategy();
+		RestartBackoffTimeStrategy restartStrategy = executionFailureHandler.getRestartBackoffTimeStrategy();
+
+		Set<ExecutionVertexID> tasksToBeRestarted = new HashSet<>();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java]: same as below, let's add at least 1 value.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	/** Reason why the failure is not recoverable. */
+	private Throwable error;
+
+	/**
+	 * Creates a result of a set of tasks to restart to recover from the failure.
+	 *
+	 * @param verticesToRestart containing task vertices to restart to recover from the failure
+	 * @param restartDelayMS indicate a delay before conducting the restart
+	 */
+	public FailureHandlingResult(Set<ExecutionVertexID> verticesToRestart, long restartDelayMS) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: when you have several constructors that heavily affect the semantics of the instance it is usually more readable to have static factory methods, like
```
public static FailureHandlingResult restartable(Set<ExecutionVertexID> verticesToRestart, long restartDelayMS)
public static FailureHandlingResult unrecoverable(Throwable error)
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * This handler deals with task failures to return a {@link FailureHandlingResult} which contains tasks
+ * to restart to recover from failures.
+ */
+public class ExecutionFailureHandler {
+
+	/** Strategy to judge which tasks should be restarted. */
+	private FailoverStrategy failoverStrategy;
```
Comment (eaglewatcherwb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: can be final

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+	/** Strategy to judge whether and when a restarting should be done. */
+	private RestartBackoffTimeStrategy restartBackoffTimeStrategy;
```
Comment (eaglewatcherwb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: same as above

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	/**
+	 * Creates the handler to deal with task failures.
+	 *
+	 * @param topology of tasks of current job
+	 * @param failoverStrategyFactory is the factory to create the failover strategy
+	 * @param restartStrategyFactory is the factory to create the restart strategy
+	 */
+	public ExecutionFailureHandler(
+		FailoverTopology topology,
+		FailoverStrategy.Factory failoverStrategyFactory,
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: I'd pass in the actual instances instead of the factories. Then we wouldn't even need to explicitly pass the topology.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+public class FailureHandlingResult {
+
+	/** Task vertices to restart to recover from the failure. */
+	private Set<ExecutionVertexID> verticesToRestart;
+
+	/** Delay before the restarting can be conducted. */
+	private long restartDelayMS;
+
+	/** Reason why the failure is not recoverable. */
+	private Throwable error;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: these fields should be final, set them to null in one of the constructors

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverStrategy.java [hunk 10]
```diff
@@ -34,4 +34,22 @@
 	 * @return set of IDs of vertices to restart
 	 */
 	Set<ExecutionVertexID> getTasksNeedingRestart(ExecutionVertexID executionVertexId, Throwable cause);
+
+	// ------------------------------------------------------------------------
+	//  factory
+	// ------------------------------------------------------------------------
+
+	/**
+	 * The factory to Instantiate {@link FailoverStrategy}.
```
Comment (eaglewatcherwb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverStrategy.java]: Factory interface for {@link FailoverStrategy}.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverStrategy.java [hunk 11]
```diff
@@ -34,4 +34,22 @@
 	 * @return set of IDs of vertices to restart
 	 */
 	Set<ExecutionVertexID> getTasksNeedingRestart(ExecutionVertexID executionVertexId, Throwable cause);
+
+	// ------------------------------------------------------------------------
+	//  factory
+	// ------------------------------------------------------------------------
```
Comment (eaglewatcherwb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverStrategy.java]: position marker is not a good choice (see Clean Code by Robert C Martin, page 67, Chapter 4).

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	/**
+	 * Returns the tasks to restart.
+	 *
+	 * @return the tasks to restart
+	 */
+	public Set<ExecutionVertexID> getVerticesToRestart() {
+		if (canRestart()) {
+			return verticesToRestart;
```
Comment (eaglewatcherwb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: return Collections.unmodifiableSet(verticesToRestart);

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartBackoffTimeStrategy.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+	/**
+	 * Notify the strategy about the task failure cause.
+	 *
+	 * @param cause of the task failure
+	 */
+	void notifyFailure(Throwable cause);
+
+	// ------------------------------------------------------------------------
+	//  factory
+	// ------------------------------------------------------------------------
```
Comment (eaglewatcherwb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartBackoffTimeStrategy.java]: see the comment in FailoverStrategy.java

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+/**
+ * Tests for {@link ExecutionFailureHandler}.
+ */
+public class ExecutionFailureHandlerTest extends TestLogger {
+
+	/**
+	 * Tests the case that task restarting is accepted.
+	 */
+	@Test
+	public void testNormalFailureHandling() throws Exception {
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java]: Exception is not thrown.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+		assertEquals(restartDelayMs, result.getRestartDelayMS());
+		assertEquals(tasksToBeRestarted, result.getVerticesToRestart());
+		assertNull(result.getError());
+	}
+
+	/**
+	 * Tests the case that task restarting is suppressed.
+	 */
+	@Test
+	public void testRestartingSuppressedFailureHandlingResult() throws Exception {
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java]: Exception is not thrown.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+		} catch (IllegalStateException ex) {
+			// expected
+		}
+	}
+
+	/**
+	 * Tests the case that the failure is non-recoverable type.
+	 */
+	@Test
+	public void testNonRecoverableFailureHandlingResult() throws Exception {
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java]: Exception is not thrown.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Tests for {@link FailureHandlingResult}.
+ */
+public class FailureHandlingResultTest extends TestLogger {
+
+	/**
+	 * Tests normal FailureHandlingResult.
+	 */
+	@Test
+	public void testNormalFailureHandlingResult() throws Exception {
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java]: same as above.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+		assertEquals(delay, result.getRestartDelayMS());
+		assertEquals(tasks, result.getVerticesToRestart());
+		assertNull(result.getError());
+	}
+
+	/**
+	 * Tests FailureHandlingResult which suppresses restarts.
+	 */
+	@Test
+	public void testRestartingSuppressedFailureHandlingResult() throws Exception {
```
Comment (eaglewatcherwb) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java]: same as above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverStrategy.java (hunk 11 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailoverStrategy.java]: Sorry but I don't have that book.
Let's just keep it as this marker usage is widely used in Flink.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java (hunk 12 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: I think it's not necessary to add extra cost to do this conversion.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java (hunk 5 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: Good idea.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java (hunk 8 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: Fine.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java (hunk 2 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: Ok.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java (hunk 12 above)
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: Instances of this class can be made immutable and I don't see a reason not to.

Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: Ok. Let's make it immutable.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java (hunk 3 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java]: OK.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java (hunk 1 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandlerTest.java]: Ok.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java (hunk 12 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: nit: we could make the `verticesToRestart` field immutable instead and return it without additional wrappers.

Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: Ok.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java]: You missed this one :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+	/**
+	 * Return result of failure handling. Can be a set of task vertices to restart
+	 * and a delay of the restarting. Or that the failure is not recoverable and the reason for it.
+	 *
+	 * @param failedTask is the ID of the failed task vertex
+	 * @param cause of the task failure
+	 * @return result of the failure handling
+	 */
+	public FailureHandlingResult getFailureHandlingResult(ExecutionVertexID failedTask, Throwable cause) {
+		if (ThrowableClassifier.getThrowableType(cause) == ThrowableType.NonRecoverableError) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: according to the [scheduler interface draft](https://docs.google.com/document/d/1fstkML72YBO1tGD_dmG2rwvd9bklhRVauh4FSsDDwXU/edit#), the `RestartBackoffTimeStrategy` is responsible for this decision.

```The delay is computed by a RestartBackoffTimeStrategy that is also responsible for deciding whether a restart should be performed at all (e.g., if the failure cause is SuppressRestartsException).```

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+		return error == null;
+	}
+
+	/**
+	 * Returns reason why the restarting cannot be conducted.
+	 *
+	 * @return reason why the restarting cannot be conducted
+	 */
+	public Throwable getError() {
+		return error;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: shouldn't this also throw an `IllegalStateException` if `error == null` like `getRestartDelayMS` and `getVerticesToRestart`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java (hunk 19 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExecutionFailureHandler.java]: Yes the design doc proposed to check the error type in the restart strategy. But that requires every strategy implementation to do the same check.

Therefore I think it's better we let the restart strategy do it's dedicated check only. 

Besides, I'm also thinking that whether we can have a composed restart strategy, which can do multiple checks for failing max limit, failing rate, etc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java (hunk 20 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java]: You are right.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java (hunk 3 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResultTest.java]: My fault.
